Mathlib/NumberTheory/NumberField/House.lean

```diff
@@ -0,0 +1,308 @@
+      apply mul_le_mul_of_nonneg_left ?_ (c_nonneg K)
+      · apply mul_le_mul_of_nonneg_right (habs kr.1 lu.1) ?_
+        · exact norm_nonneg ((canonicalEmbedding K) ((algebraMap (𝓞 K) K)
+            ((newBasis K) lu.2)))
+    ·  apply mul_le_mul_of_nonneg_left ?_ (mul_nonneg (c_nonneg K) Apos)
+       · simp only [supOfBasis, le_sup'_iff, mem_univ]; use lu.2
+    · rw [mul_right_comm]
+      exact mul_le_mul_of_nonneg_right
+        (mul_le_mul_of_nonneg_right (le_max_right ..) (supOfBasis_nonneg K)) Apos
+  · rw [mul_nonneg_iff]; left; exact ⟨c₂_nonneg K, Apos⟩
+
+/-- `c₁ K` is the product of `finrank ℚ K` and  `c₂ K` and depends on `K`. -/
+private def c₁ := finrank ℚ K * c₂ K
+
+private theorem house_le_bound : ∀ l, house (ξ K x l).1 ≤ (c₁ K) *
+    ((c₁ K * q * A)^((p : ℝ) / (q - p))) := by
+  let h := finrank ℚ K
+  intros l
+  calc _ = house (algebraMap (𝓞 K) K (∑ r, (x (l, r)) * ((newBasis K) r))) := rfl
+       _ ≤ ∑ r, house (((algebraMap (𝓞 K) K) (x (l, r))) *
+        ((algebraMap (𝓞 K) K) ((newBasis K) r))) := ?_
+       _ ≤ ∑ r, ‖x (l,r)‖ * house ((algebraMap (𝓞 K) K) ((newBasis K) r)) := ?_
+       _ ≤ ∑ r, ‖x (l, r)‖ * (supOfBasis K) := ?_
+       _ ≤ ∑ _r : K →+* ℂ, ((↑q * h * ‖asiegel K a‖) ^ ((p : ℝ) / (q - p))) * supOfBasis K := ?_
+       _ ≤ h * (c₂ K) * ((q * c₁ K * A) ^ ((p : ℝ) / (q - p))) := ?_
+       _ ≤ c₁ K * ((c₁ K * ↑q * A) ^ ((p : ℝ) / (q - p))) := ?_
+  · simp_rw [← _root_.map_mul, map_sum]; apply house_sum_le_sum_house
+  · apply sum_le_sum; intros r _; convert house_mul_le ..
+    simp only [map_intCast, house_intCast, Int.cast_abs, Int.norm_eq_abs]
+  · apply sum_le_sum; intros r _; unfold supOfBasis
+    apply mul_le_mul_of_nonneg_left ?_ (norm_nonneg (x (l,r)))
+    · simp only [le_sup'_iff, mem_univ, true_and]; use r
+  · apply sum_le_sum; intros r _
+    apply mul_le_mul_of_nonneg_right ?_ (supOfBasis_nonneg K)
+    exact le_trans (norm_le_pi_norm x ⟨l, r⟩) hxbound
+  · simp only [Nat.cast_mul, sum_const, card_univ, nsmul_eq_mul]
+    rw [Embeddings.card, mul_comm _ (supOfBasis K), c₂, c₁, ← mul_assoc]
+    apply mul_le_mul
+    · apply mul_le_mul_of_nonneg_left ?_ (Nat.cast_nonneg' _)
+      · exact le_mul_of_one_le_left (supOfBasis_nonneg K) (le_max_left ..)
+    · apply Real.rpow_le_rpow (mul_nonneg (mul_nonneg (Nat.cast_nonneg' _) (Nat.cast_nonneg' _))
+        (norm_nonneg _))
+      · rw [← mul_assoc, mul_assoc (_*_)]
+        apply mul_le_mul_of_nonneg_left (asiegel_remark K a habs Apos)
+          (mul_nonneg (Nat.cast_nonneg' _) (Nat.cast_nonneg _))
+      · exact div_nonneg (Nat.cast_nonneg' _) (sub_nonneg.2 (mod_cast hpq.le))
+    · apply Real.rpow_nonneg
+      exact mul_nonneg (mul_nonneg (Nat.cast_nonneg' _) (Nat.cast_nonneg' _))
+        (norm_nonneg _)
+    · exact mul_nonneg (Nat.cast_nonneg' _) (mul_nonneg (le_trans zero_le_one (le_max_left ..))
+        (supOfBasis_nonneg _))
+  · rw [mul_comm (q : ℝ) (c₁ K)]; rfl
+
+theorem exists_ne_zero_int_vec_house_le :
```

Comment:
Can you add a docstring to this theorem? Also adding it to the beginning of the file if you think it is the main result.